### ISSUE TITLE
QUESTION: Why the store's dispatch does not return the new state?

### ISSUE THREAD

**Issue by mictian** [onerzafer/rxdx#1]
Hi,

First of all, great 3 [articles](https://hackernoon.com/diy-redux-with-rxjs-rxdx-23163a87ade1).

In your middlewares you expect, at least that is what seems by reading the code, that the next (dispatch) function to return the new state. Or at least something.
In the [devMiddleware](https://github.com/onerzafer/rxdx/blob/master/src/devToolsMiddleware.js#L22) and in the [effectsMiddleware](https://github.com/onerzafer/rxdx/blob/master/src/effectsMiddleware.js#L17) you save in a variable and even return the result of invoking next. But in practice it is just `undefined`.
In the [store](https://github.com/onerzafer/rxdx/blob/master/src/store.js#L14) you're not returning anything.

I just wonder why it is like that.

Regards,
Mictian

**Comment by onerzafer** [comment]
Thanks @mictian 
Regarding your question, I would like to clarify one point. Store.dispatch is the final destination of the middleware chain and its responsibility is updating the state. Since only reducers operate on state and middlewares on the action I would say that middlewares returning an action which may or may not be the modified version of the original one. The most simplistic explanation can be done like:
```
dispatch(middleware(action))
```

**Comment by mictian** [comment]
Hi @onerzafer,

That makes all the sense. But then, why are you returning and saving into a variable 'undefined'  in your middlewares: [devMiddleware L22](https://github.com/onerzafer/rxdx/blob/master/src/devToolsMiddleware.js#L22) and in the [effectsMiddleware L17](https://github.com/onerzafer/rxdx/blob/master/src/effectsMiddleware.js#L17) ?

Regards,
Mictian

**Comment by onerzafer** [comment]
Hi @mictian 
It not undefined its an action manipulated by another middleware
```
const result = next(action); // next(action) returns an action and next itself is a middleware
```
So maybe the naming is confusing?

Best

**Comment by mictian** [comment]
Hi @onerzafer 

Thanks for the explanation. My confusion was that if I only use one middleware `next` would be the original `store.dispatch`.
Actually, the _first_ invoked middleware will have store.dispatch as next, right?

Regards,
Mictian

**Comment by onerzafer** [comment]
Hi @mictian 

Nope! On the contrary, dispatch has the first middleware as its action

Best

**Comment by mictian** [comment]
Hi @onerzafer,

Heheh! I dont know if I am understanding it more or less :P.

I cannot see where I am misunderstanding the code.

I start with the following:
- Middlewares have the following signature: `(store) => (next) => (action) => state`
- Apply middlware function has the following signature: `(...middlewares) => (()=> store) => (args) => store`
- Create sore has the following one: `(reducer, initialState, enhancer) => store`

I hope that this basic bits I have it clear.

So, my understanding of the initialization (application startup, store-creation, or whatever is called) is the following:
1. Create store, this is done [here](https://github.com/onerzafer/rxdx/blob/master/src/store.js#L36). Simple and straighforward.
2. The store is enhaced by intercepting all dispatched actions and let middlewares run before actions are really "executed". This is done [here](https://github.com/onerzafer/rxdx/blob/master/src/store.js#L37)
3. Finally the initial action '@@store INIT' is dispatch and the store returned.

I again hope this is ok, as it is just a literal exaplanation of the creation store function.

The part I am interested about is the second point, enhance the store.   
In order to enhace the store an "adapter" or fuction that can "enhance" the store by combine it with all the middlewares is needed. And in fact this is the work done by [applyMiddleware](https://github.com/onerzafer/rxdx/blob/master/src/applyMiddleware.js).

applyMiddleware, I THINK, works in the following way:
- First, it accepts all middlwares to be applied. An example can be seen [here](https://github.com/onerzafer/rxdx-sample-app/blob/master/src/stores/store.js#L16).  
_Note_ that as we started with the following signature for the applyMiddleware `(...middlewares) => (()=> store) => (args) => store`, now we have the following one `(()=> store) => (args) => store`.
- Second, it returns a function that accept a create store functiom as parameter. This piece of code can be seen in [applyMiddleware L2](https://github.com/onerzafer/rxdx/blob/master/src/applyMiddleware.js#L2).
- This function is invoke by createStore function in case the enhancer passed-in is truly, I mean [here](https://github.com/onerzafer/rxdx/blob/master/src/store.js#L37), in this piece of code `enhancer(() => store)`.
- Once the createStore function invokes the enhancer, applyMiddelware will return another function. In this case the one given by the following piece of code: `(...args) => {` in line 2 of the applyMiddleware file.  
_Note_ that from the previous signature `(()=> store) => (args) => store` now we have only the following one `(args) => store` from the applyMiddleware.
- Now again, the createStore function invoke this new function by passing the reducer and the initialState as parameters. This can be seen a in the line 37 of the store file with the following code: `enhancer(() => store)(reducer, initialState)`.
- The previous step means that the `...args` from the applyMiddelware (I the args from the line 2 in the applyMiddleware file) will be `reducer` and `initialState`.   
Those params are used to [retrieve the store](https://github.com/onerzafer/rxdx/blob/master/src/applyMiddleware.js#L3). (Note aside, those param are not used as the createSore function is parameterless and only returns the already created store).   
Now, the part where the middlewares are process become. If I am not mistaken the signature for middleware is the following: `(store) => (next) => (action) => state`

1. In the applyMiddleware first all middlewares are "initialized" by iterating over all them and invoke this first function. This case be seen [here](https://github.com/onerzafer/rxdx/blob/master/src/applyMiddleware.js#L4-L5).   
_Note_ that will consume that param from the middleware signature. From `(store) => (next) => (action) => state` now we have `(next) => (action) => state` (IF I am not missing anything)
2. The result of the previous map is an array of functions with the previous shown signature.   
From this result applyMiddlware reduce all those function into one single function with exactly the same signature: `(next) => (action) => state`.   
I think here is my confusion, `next` in fact is a function where its signarure I think should be `(action) => state` or at least it should accept an _action_.   
The way all middlewares are combined is by creating a new function: `(...args) => a(b(...args))` that accepts any argument and when invoked will invoke the first middlewares with those params and use the result from the first middleware' result as the first param for the second middleware (and so on and so on).   
In other words, the reduce will return the following signature: `(...args) => (action) => state`. In details, if we have two middlwares `[(next) => (action) => state, (next) => (action) => state]`, the first step to reduce them will be creating the following function: `(args) => ((next) => (action) => state)(args)` which is a function that accept `args` and immediately invoke the middleware, which will return another function with the signature `(action) => state`. This result will then be used as first parameter for the second middleware (reason why I think that the next function should accept an action and return a new state).   
Finally, this function is invoked by passing-in the original store.dispatch. This can be seen in line 6 of the [applyMiddleware](https://github.com/onerzafer/rxdx/blob/master/src/applyMiddleware.js#L6) with the following piece of code `(store.dispatch.bind(store))`.  
Here it is what leads me to think that the first middleware being execute will recevie as next the original dispatch.   

Besides, that points also leads me to think that, as the original store.dispatch does not return any sate, why you was saving in the middlewares the result of invoking next.

Please let me know where I mistkaken.

Best Regards,   
Mictian

**Comment by onerzafer** [comment]
Hi @mictian 

Thanks for your effort I appreciate it very much!

I spend some time on what we are discussing and how do we need to approach the discussion. 
First of all, I created an isolated sample implementation and I saw you were right since the beginning.
[https://codesandbox.io/s/o798jmpm9q?fontsize=14](https://codesandbox.io/s/o798jmpm9q?fontsize=14)
Observe the logs you will see that you are right.

So I was thinking that the discussion was way too long and inefficient. If you don't mind, I have a couple of suggestions for both of us to improve it.

- Always provide working, isolated examples to support your ideas.
- Consider contributing Instead of proving yourself on comments!
- Allways write test  (this was for me)

So since you pointed the problem would you like to create a pull request?

Bests

**Comment by mictian** [comment]
Hi @onerzafer ,

Thank you for your answer.
Personally, I don't care who is right or wrong, but to learn a bit more every time.
I did not made any code as I was genuinely doubting on why it was working on certain way. Thanks to your answers I realized that the first invocation was using the original dispatcher and other will use the result from the first middleware.

I do will create a PR 😄.

Again, thank you for all the feedback.

Regards,
Mictian
